### DISCUSSION TITLE
Returned feature coordinates different then original feature?

### DISCUSSION THREAD

**Discussion by bdullemond** [maplibre/maplibre-gl-js#609] [Q&A] [answered]
Hi there, loving maplibre for the past 6 months in a project. You are doing great work. 

I ran into something and I want to confirm its expected behaviour. I have a map with points and when I click on those points I have a popup appear to display the clicked feature's details include coordinate. However, if I make a point at say [-75.0, 45.0] depending on the zoom level I get an approximation of that coordinate in the feature's geometry property, with it being more accurate as I zoom in.
![image](https://user-images.githubusercontent.com/411881/140056572-fdd34f74-4727-40be-a9ef-eea84bd13ff4.png)
The same thing has happened with map.queryRenderedFeatures(...) call.

Is this expected? I really want to get the original coordinates the feature was created with.
Thanks!

EDIT: example: https://codepen.io/Terribill/pen/KKvZNjx

**Comment by wipfli** [comment]
I am glad you enjoy using MapLibre. From what I have seen in the unit test, numerical equality checks are usually restricted to only a few digits. So I guess what you show here is the expected behavior.

**Comment by astridx** [comment, marked as answer]
I also think that the output in your popup is correct. If I read the [documentation ](https://maplibre.org/maplibre-gl-js-docs/api/map/#map.event:mouseenter) and the code for `map.on('mouseenter', 'places',..)` correctly, then the coordinate of the mouse position on the `places ` `layer `is displayed as **accurately** as possible. 

The `source`, I mean the value for `e.features[0].geometry.coordinates` of the `GeoJson` is not the target of the event. I think that is what you expect. But you would have to do this differently. 

In your case, however, I think that the easiest thing to do is to use:

`popup.setLngLat(coordinates).setHTML(description + coordinates[0].toFixed(2) + "," + coordinates[1].toFixed(2)).addTo(map);`

**Comment by bdullemond** [comment]
Thanks for the replies. Unfortunately I need 6 decimal places so I found a workaround of putting the source coordinates in the feature properties for future use in popups and that works well.

**Comment by cooljacob204** [comment]
I have a somewhat similar issue. A valid polygon (no self intersections, etc...) that I provided became invalid due to a self intersection and failed validation because the coordinates returned are slightly different from the geometry in the onClick event. Seeing differences up to the 5th digit in some of my tests.

My project renders a floor (walls, doors, chairs, desks) on the map so it needs to be fairly accurate.
